maup/repair.py

```diff
@@ -1,27 +1,1217 @@
+    then by largest shared perimeter.
+
+
+Brief sketch of the new gap-closing algorithm:
+
+(1) For any hole whose boundary touches 3 or fewer polygons, attach it to the
+    polygon with which it shares the greatest perimeter.  (Doesn't really matter;
+    in practice holes have very small area, and adjacency relationships won't be
+    affected regardless of where we attach it.)
+
+(2) For a hole whose boundary touches 4 or more polygons, find the pair of 
+    non-adjacent polygons with the shortest distance between them.  Cut out a
+    piece of the hole and use it to attach these two polygons with a boundary
+    segment of positive length.  (This is a "best guess" for when two polygons
+    on opposite sides of a long, thin gap should really be adjacent.)
+    This leaves behind two new holes, each with strictly fewer boundary
+    polygons than the original.
+
+(3) Repeat until all holes are filled.
+
+
+Brief sketch of the new rook-to-queen algorithm:
+
+Let L be a linestring with length below the specified threshold.
+
+(1) Construct a small disk D that contains the linestring within its interior.
+    Find all polygons P_1, ..., P_k that intersect D.  
+
+(2) Take the unary_union of all the boundaries of P_1, ..., P_k, and D, and
+    polygonize this structure.  Assign all pieces outside of D to the polygons
+    that they came from, and assign the pieces inside D to D.  (This is 
+    equivalent to taking the difference of each polygon minus D, except that
+    it works cleanly and reliably!)
+
+(3) For each of the polygons whose boundary (now) contains an arc of D's boundary
+    circle, create a "pie wedge" inside D subtending this arc, and join it to    
+    the polygon that it shares a boundary with.
+    
 """
-Some of these functions are based on the functions in Mary Barker's
-check_shapefile_connectivity.py script in @gerrymandr/Preprocessing.
+
+
+# Some useful short functions:
+
+# Count the number of connected components of a shapely object:
+def num_components(geom):
+    if geom.is_empty:
+        return 0
+    elif (
+        (geom.geom_type == "Polygon")
+        or (geom.geom_type == "Point")
+        or (geom.geom_type == "LineString")
+    ):
+        return 1
+    elif (
+        (geom.geom_type == "MultiPolygon")
+        or (geom.geom_type == "MultiLineString")
+        or (geom.geom_type == "GeometryCollection")
+    ):
+        return len(geom.geoms)
+
+
 """
+MAIN FUNCTION
+"""
+
+
+def autorepair(
+    geometries0_df, fill_gaps=True, min_rook_length=None, holes_to_keep=None
+):
+
+    if isinstance(geometries0_df, geopandas.GeoSeries):
+        geometries_df = geopandas.GeoDataFrame(geometry=geometries0_df)
+
+    elif isinstance(geometries0_df, geopandas.GeoDataFrame):
+        geometries_df = geometries0_df.copy()
+
+    else:
+        raise TypeError(f"Input must be a geopandas GeoSeries or GeoDataFrame!")
+
+    geometries_df = geometries0_df.copy()
+
+    # Ensure that geometries are 2-D and not 3-D!
```

Comment:
This loop feels like it could be expensive (though I admittedly haven't benchmarked)—we're serializing and deserializing every polygon in the input DataFrame, regardless of whether it's necessary. If someone's passing in a 3D geometry, that's quite weird, and I don't think we should handle that by suppressing one of the dimensions. What if we looped through all of the geometries, checked their dimension, and raised a `ValueError` if we found any 3D geometries?